examples/factory/tests/tests.rs

```diff
@@ -0,0 +1,256 @@
+    product::{Product, ProductState},
+};
+
+/// The test accounts.
+const ALICE: AccountAddress = AccountAddress([0; 32]);
+const BOB: AccountAddress = AccountAddress([1; 32]);
+
+const SIGNER: Signer = Signer::with_one_key();
+const ACC_INITIAL_BALANCE: Amount = Amount::from_ccd(10000);
+
+/// Test the case of producing two products from one factory.
+#[test]
+fn test_success() {
+    let (mut chain, module_ref) = initialize_chain();
+    // Set up the factory instance.
+    let factory = create_factory(&mut chain, module_ref);
+
+    // Create a product instance.
+    let product = create_product(&mut chain, module_ref, ALICE);
+    // Check that it is uninitialized
+    let prod_state = view_product(&chain, product).expect("Get view of uninitialized product");
+    assert_eq!(prod_state, ProductState::Uninitialized);
+    // "Produce" the product with the factory.
+    call_produce(&mut chain, factory, product, ALICE).expect("Produce successful");
+    // Check that it is initialized as we expect.
+    let prod_state = view_product(&chain, product).expect("Get view of initialized product");
+    assert_eq!(
+        prod_state,
+        ProductState::Initialized(Product {
+            factory,
+            index: 0,
+        })
+    );
+
+    // Create and produce a second product (from a different account), and check the
+    // result.
+    let product2 = create_product(&mut chain, module_ref, BOB);
+    call_produce(&mut chain, factory, product2, BOB).expect("Produce successful");
+    let prod_state = view_product(&chain, product2).expect("Get view of initialized product");
+    assert_eq!(
+        prod_state,
+        ProductState::Initialized(Product {
+            factory,
+            index: 1,
+        })
+    );
+
+    // Check that the factory records having 2 products, and that they match what we
+    // expect.
+    assert_eq!(view_product_count(&chain, factory).expect("View product count successful"), 2);
+    assert_eq!(view_product_at(&chain, factory, 0).expect("View of product at index"), product);
+    assert_eq!(view_product_at(&chain, factory, 1).expect("View of product at index"), product2);
+}
+
+/// Test that a product instance cannot be produced using an account that did
+/// not create the product instance.
+#[test]
+fn test_not_authorized() {
+    let (mut chain, module_ref) = initialize_chain();
+    let factory = create_factory(&mut chain, module_ref);
+    // ALICE creates the product instance, but BOB invokes produce on the factory.
+    // This should fail.
+    let product = create_product(&mut chain, module_ref, ALICE);
+    let err = call_produce(&mut chain, factory, product, BOB).expect_err(
+        "Should fail when product is called from a different account than the creator of the \
+         product.",
+    );
+    assert_eq!(
+        err.reject_code(),
+        Some(Reject::from(FactoryError::InitializeFailed).error_code.get()),
+        "Reject code should be InitializeFailed"
+    );
```

Comment:
Just for info: This is how we check the error in our other examples. Not feeling strongly about either way. 

    // Check that the correct error is returned.
    let rv: ContractError = update.parse_return_value().expect("ContractError return value");
    assert_eq!(rv, ContractError::Unauthorized);
    
https://github.com/Concordium/concordium-rust-smart-contracts/blob/main/examples/cis2-multi/tests/tests.rs#L235C1-L237C49